### ISSUE TITLE
Postgres planner produce incorrect plan with CTE with volatile target list for case of Segment General and General locuses

### ISSUE THREAD

**Issue by KnightMurloc** [greenplum-db/gpdb#16785]
## Bug Report
If the CTE with General or Segment General locus types contain a volatile function in the target list, then such CTE should be brought to SingleQE since the General and Segment General locus should give the same result on each of the segments, whereas a volatile function can give a different result on each call. And query can produce inccorect result. A similar situation is [handled](https://github.com/greenplum-db/gpdb/blob/aa5d4fc5e49e7781fe4da70f68b92261f91274d5/src/backend/optimizer/path/allpaths.c#L1720-L1733) for subqueries but not for CTE, union/except/intersect and multiset functions.

I have a patch (https://github.com/arenadata/gpdb/commit/120aaef599cf4cce1fac430b833a1b74b4a5d8bd) that fixes this problem by moving the processing of volatile functions in the target list for plan nodes with the General and Segment General locus type to `subquery_planner` so that processing is performed for all subqueries.

### Greenplum version or build
`PostgreSQL 9.4.26 (Greenplum Database 6.26.0+dev.8.gf451b4d1fe build dev) on x86_64-unknown-linux-gnu, compiled by gcc (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0, 64-bit compiled on Nov 23 2023 15:22:55 (with assert checking)`
### OS version and uname -a
Ubuntu 22.04 LTS (docker)
`Linux MURLOC 6.1.55-1-MANJARO #1 SMP PREEMPT_DYNAMIC Sat Sep 23 12:13:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`
### autoconf options used ( config.status --config )
`'--enable-debug-extensions' '--with-python' '--disable-gpcloud' '--enable-depend' '--enable-cassert' 'CFLAGS=-O0 -g3 -ggdb -fno-omit-frame-pointer -Wno-error=deprecated-copy' 'CXXFLAGS=-O0 -g3 -fno-omit-frame-pointer -Wno-error=deprecated-copy' '--prefix=/gpdb/gpdb_bin' '--with-libxml'`
### Installation information ( pg_config )
```
BINDIR = /gpdb/gpdb_bin/bin
DOCDIR = /gpdb/gpdb_bin/share/doc/postgresql
HTMLDIR = /gpdb/gpdb_bin/share/doc/postgresql
INCLUDEDIR = /gpdb/gpdb_bin/include
PKGINCLUDEDIR = /gpdb/gpdb_bin/include/postgresql
INCLUDEDIR-SERVER = /gpdb/gpdb_bin/include/postgresql/server
LIBDIR = /gpdb/gpdb_bin/lib
PKGLIBDIR = /gpdb/gpdb_bin/lib/postgresql
LOCALEDIR = /gpdb/gpdb_bin/share/locale
MANDIR = /gpdb/gpdb_bin/share/man
SHAREDIR = /gpdb/gpdb_bin/share/postgresql
SYSCONFDIR = /gpdb/gpdb_bin/etc/postgresql
PGXS = /gpdb/gpdb_bin/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--enable-debug-extensions' '--with-python' '--disable-gpcloud' '--enable-depend' '--enable-cassert' 'CFLAGS=-O0 -g3 -ggdb -fno-omit-frame-pointer -Wno-error=deprecated-copy' 'CXXFLAGS=-O0 -g3 -fno-omit-frame-pointer -Wno-error=deprecated-copy' '--prefix=/gpdb/gpdb_bin' '--with-libxml'
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/include/libxml2 -I/gpdb/gpdb_bin/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -Werror=implicit-fallthrough=3 -Wno-format-truncation -Wno-stringop-truncation -O0 -g3 -ggdb -fno-omit-frame-pointer -Wno-error=deprecated-copy -std=gnu99 -Werror=uninitialized -Werror=implicit-function-declaration -I/gpdb/gpdb_bin/include
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/gpdb/gpdb_bin/lib',--enable-new-dtags -L/gpdb/gpdb_bin/lib
LDFLAGS_EX = 
LDFLAGS_SL = 
LIBS = -lpgcommon -lpgport -lxerces-c -lxml2 -lrt -lzstd -lcrypt -lm -L/gpdb/gpdb_bin/lib
VERSION = PostgreSQL 9.4.26
```
### Expected behavior
The same values in the 'rand' column.
### Actual behavior
```
       rand        
-------------------
 0.150738566648215
 0.150738566648215
 0.150738566648215
 0.150738566648215
 0.150738566648215
 0.150738566648215
 0.150738566648215
 0.150738566648215
 0.332179193384945
 0.332179193384945
 0.332179193384945
 0.332179193384945
 0.332179193384945
 0.332179193384945
 0.332179193384945
 0.332179193384945
 0.979839109349996
 0.979839109349996
 0.979839109349996
 0.979839109349996
 0.979839109349996
(21 rows)
```
### Step to reproduce the behavior
```
set optimizer = off;
create table r (a int, b int) distributed replicated;
create table d (b int, a int default 1) distributed by (b);

insert into d select * from generate_series(0, 20) j;
alter table d set distributed randomly;
insert into r values (1, 1), (2, 2), (3, 3);

-- Segment General
with cte as (
    select a, b * random() as rand from r
)
select rand from cte join d on cte.a = d.a;

-- General
with cte as (
    select a as a, a * random() as rand from generate_series(0, 3)a
)
select rand from cte join d on cte.a = d.a;
```

**Comment by Aegeaner** [comment]
@KnightMurloc The idea is quite like the one in this PR: https://github.com/greenplum-db/gpdb/pull/10418
Can we handle the CTE issue using `turn_volatile_seggen_to_singleqe()` method introduced by above PR?

**Comment by KnightMurloc** [comment]
yes, for GP7, the CTE problem can be solved using `turn_volatile_seggen_to_singleqe` in the same way as for subquery, but what about GP6?

**Comment by Aegeaner** [comment]
> yes, for GP7, the CTE problem can be solved using `turn_volatile_seggen_to_singleqe` in the same way as for subquery, but what about GP6?

@KnightMurloc 
The introduced `turn_volatile_seggen_to_singleqe()` PR have been backported to the 6X branch:
https://github.com/greenplum-db/gpdb/pull/10495

And `set_cte_pathlist()` should have call it already, but if it is not working, we should figure out why:
![截图 2023-11-30 18-51-11](https://github.com/greenplum-db/gpdb/assets/1671185/b64822f4-0859-4eee-b3ee-dacb5439b43b)

**Comment by KnightMurloc** [comment]
it looks like we only check baserestrictinfo for volatile functions, but not reltargetlist. https://github.com/greenplum-db/gpdb/blob/08d747e805904904d386986dc4d8ea264514e5c8/src/backend/optimizer/path/allpaths.c#L518-L520

**Comment by Aegeaner** [comment]
> it looks like we only check baserestrictinfo for volatile functions, but not reltargetlist.
> 
> https://github.com/greenplum-db/gpdb/blob/08d747e805904904d386986dc4d8ea264514e5c8/src/backend/optimizer/path/allpaths.c#L518-L520
> 
> Upd: `reltargetlist` does not contain information about functions. but this information is in the `root->parse->targetList`.

Thanks, I am working on a draft PR trying to fix this issue: https://github.com/greenplum-db/gpdb/pull/16798

But it seems not easy to use `turn_volatile_seggen_to_singleqe()` to replace all the similar processing scenes in the 6X branch, because we do need to manipulate with the `Plan` node instead of the `Path` node, without overcomming the big refactoring gap between GP6 & GP7.

**Comment by KnightMurloc** [comment]
it looks like with this solution, we have additional Result nodes in some plans with a volatile target list. it also looks like we no longer need the condition for the target list here:
https://github.com/greenplum-db/gpdb/blob/84e28613dddb0be3d4525fc73feb6c3937697a51/src/backend/optimizer/path/allpaths.c#L1720-L1733
https://github.com/greenplum-db/gpdb/blob/84e28613dddb0be3d4525fc73feb6c3937697a51/src/backend/optimizer/plan/subselect.c#L670-L677
